### PR TITLE
Feature/logging errors

### PR DESCRIPTION
**Pull request by dgraeber**
*Issue #, if available:*
#42 
#29

*Description of changes:*
Adding support for custom error classes from codeseeder, optimized logging and log outputs, code refactoring for extensibility

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: seedfarmer/__main__.py [hunk 1]
```diff
@@ -348,14 +356,23 @@ def list_module_metadata(
...
         enable_debug(format=DEBUG_LOGGING_FORMAT)
-    _logger.debug(f"We are getting module data for {module} in {group} of {deployment}")
+    if _logger.isEnabledFor(logging.DEBUG):
+        _logger.debug("We are getting module data for  %s in %s of %s", module, group, deployment)
 
-    val = json.dumps(mi.get_module_metadata(deployment, group, module))
-    sys.stdout.write(val)
+    metadata_json = mi.get_module_metadata(deployment, group, module)
+    if not export_local_env:
+        sys.stdout.write(json.dumps(metadata_json))
```
Comment (chamcca) [inline seedfarmer/__main__.py:369]: why the explicit write to stdout here and below? should these be log messages?

File: seedfarmer/__main__.py [hunk 2]
```diff
@@ -203,8 +203,8 @@ def store_module_metadata(
 ) -> None:
     if debug:
         enable_debug(format=DEBUG_LOGGING_FORMAT)
-
-    _logger.debug(f"Writing metadata for module {module} in deployment {deployment}")
+    if _logger.isEnabledFor(logging.DEBUG):
```
[the remark this answers is not in the document]
Comment (dgraeber) [reply seedfarmer/__main__.py:206]: https://docs.python.org/3/howto/logging.html#optimization

File: seedfarmer/__main__.py (hunk 1 above)
Comment (dgraeber) [reply seedfarmer/__main__.py:369]: the CLI support for metadata has always been to write to std out (the deployspec uses the output to export to env).

Should we change?

Comment (chamcca) [reply seedfarmer/__main__.py:369]: got it. all good,

File: seedfarmer/__main__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chamcca) [reply seedfarmer/__main__.py:206]: that makes sense in cases where the string substitution requires a function call, like the following, where the json.dumps() function is always called on the .debug() call even if the debug message isn't going to be written.
```
        if _logger.isEnabledFor(logging.DEBUG):
            _logger.debug(
                "DeploymentManifest for deploy after filter =  %s", json.dumps(deployment_manifest_wip.dict())
            )
```
but most of the uses of .isEnableFor() don't contain a function call in their .debug(), they have variables that are already set like:
```
        _logger.debug(f"Writing metadata for module %s in deployment %s", module, deployment)
```
in these cases nothing is gained by checking .isEnabledFor()

[the remark this answers is not in the document]
Comment (dgraeber) [reply seedfarmer/__main__.py:206]: I can remove on all non-function calls....

**Review by chamcca — APPROVED** [review]

**Pull request merged**